File: Strings/Valid_palindrome.py
'''Name : Abhinav kumar
Github username : Abhinavcode13
Repository name : data-structures-and-algorithms
Problem :  Find Number of Good Pairs in Go
Issue Number : #648
Problem statement : Check whether a given string is a Valid Palindrome in Python

Explanation of the below cpp code :

The is_valid_palindrome() function is the same as the one in the previous example, which checks whether a given string is a valid palindrome or not.

The input() function is used to take input from the user.
It displays a prompt message "Enter a string:" and waits for the user to enter a value followed by pressing the Enter key. The entered value is stored in the variable s.

The if statement checks if the input string s is a valid palindrome or not using the is_valid_palindrome() function.
If it is, it prints the message "The string is a valid palindrome", otherwise it prints "The string is not a valid palindrome".

The time complexity of the given code is O(n)

'''
------------------------------------------------------------------------------------------------------//Python code begins here----------------------------------------------------------------------------------------------------------------------------------

def is_valid_palindrome(s):
    s = ''.join(filter(str.isalnum, s)).lower()
    return s == s[::-1]

s = input("Enter a string: ") #Taking input from user

if is_valid_palindrome(s):
    print("The string is a valid palindrome") #True case
else:
    print("The string is not a valid palindrome")  #False case

File: Strings/valid_palindrome.js
/*Name : Abhinav kumar
Github username : Abhinavcode13
Repository name : data-structures-and-algorithms
Problem :Two Pointers: Check whether a given string is a Valid Palindrome in Javascript
Issue Number : #649
Problem statement : 

Explanation of the below Java code :

we first define the isPalindrome() function, which converts the input string to lowercase and removes all non-alphanumeric characters using a regular expression, and then uses two pointers to loop through the string and check whether it is a valid palindrome.

We then use the prompt() function to display a popup dialog box and prompt the user to enter a string input. The input value is then stored in the str variable.

We then call the isPalindrome() function with the input string as an argument to check whether it is a palindrome. If it is, we use the console.log() function to display a message saying that the input string is a palindrome. If it is not, we display a message saying that the input string is not a palindrome.



*/

-------------------------------------------------------------------------------------------------------//Java code begins here-------------------------------------------------------------------------------------------------------------------------------

function isPalindrome(str) {
  // Convert the input string to lowercase and remove non-alphanumeric characters
  str = str.toLowerCase().replace(/[^a-z0-9]/g, '');

  // Initialize two pointers, one starting from the beginning of the string, and the other from the end
  let left = 0;
  let right = str.length - 1;

  // Loop through the string while the pointers haven't crossed each other
  while (left < right) {
    // If the characters at the left and right pointers don't match, the string is not a palindrome
    if (str[left] !== str[right]) {
      return false;
    }

    // Move the pointers towards each other
    left++;
    right--;
  }

  // If the loop finishes, the string is a palindrome
  return true;
}

// Prompt the user for a string input
let str = prompt("Enter a string:");

// Check whether the input string is a palindrome
if (isPalindrome(str)) {
  console.log(`${str} is a palindrome!`);
} else {
  console.log(`${str} is not a palindrome.`);
}
